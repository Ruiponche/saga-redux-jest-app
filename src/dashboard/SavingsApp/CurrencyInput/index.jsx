import React, { Component } from 'react'
import NumberFormat from 'react-number-format'

class CurrencyInput extends Component {
  constructor(props) {
    super(props)
    this.state = { value: undefined, formattedValue: undefined }
  }

  setValues = values => {
    console.log(values)
  }

  onValueChange = values => {
    const { onValueChange } = this.props
    this.setValues(values)
    onValueChange && onValueChange(values.floatValue)
  }

  render() {
    const { placeholder } = this.props
    return (
      <NumberFormat
        placeholder={placeholder}
        thousandSeparator={true}
        prefix={'$'}
        onValueChange={this.onValueChange}
        allowNegative={false}
      />
    )
  }
}

export default CurrencyInput
