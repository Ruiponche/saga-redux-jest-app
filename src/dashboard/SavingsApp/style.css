.savings-app-container {
  width: 60%;
  padding: 20px;
  margin-left: auto;
  margin-right: auto;
  display: grid;
  grid-gap: 20px;
  grid-template-columns: 1fr 1fr;
  grid-template-areas: 'label value';
}

.savings-app-label {
  grid-area: label;
  align-self: right;
}
