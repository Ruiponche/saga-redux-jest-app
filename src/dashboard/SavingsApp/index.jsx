import SavingsApp from './SavingsApp'

export default SavingsApp
