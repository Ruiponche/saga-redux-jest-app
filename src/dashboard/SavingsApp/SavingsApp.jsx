import React, { Component } from 'react'
import CurrencyInput from '../common/CurrencyInput'
import PercentageInput from '../common/PercentageInput'
import NumberFormat from 'react-number-format'
import Typography from '@material-ui/core/Typography'
import InputFrame from '../common/layout/InputFrame'
import FormGroup from '@material-ui/core/FormGroup'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import Switch from '@material-ui/core/Switch'
import './style.css'

class SavingsApp extends Component {
  constructor(props) {
    super(props)
    this.state = {
      // currentSavings: undefined,
      baseSalary: undefined,
      monthlyExpenses: undefined,
      annualInflation: undefined,
      annualInterestRate: undefined,
      years: undefined,
      totalSavings: undefined,
      showAnnualInflation: false,
      showMonthlyExpenses: false
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.inputHasChanged(prevState, this.state)) {
      this.setState({ totalSavings: this.totalSavings() })
    }
  }

  inputHasChanged = (prevState, currentState) => {
    if (
      (prevState.baseSalary !== currentState.baseSalary ||
        (prevState.monthlyExpenses !== currentState.monthlyExpenses &&
          currentState.showMonthlyExpenses) ||
        (prevState.annualInflation !== currentState.annualInflation &&
          currentState.showAnnualInflation) ||
        prevState.annualInterestRate !== currentState.annualInterestRate ||
        prevState.years !== currentState.years) &&
      (currentState.baseSalary &&
        (currentState.monthlyExpenses || !currentState.showMonthlyExpenses) &&
        (currentState.annualInflation || !currentState.showAnnualInflation) &&
        currentState.annualInterestRate &&
        currentState.years)
    ) {
      return true
    }
    return false
  }

  totalSavings = () => {
    const {
      currentSavings,
      baseSalary,
      monthlyExpenses,
      annualInflation,
      annualSavings,
      annualInterestRate,
      years,
      showAnnualInflation,
      showMonthlyExpenses
    } = this.state
    const annualInflationValue = showAnnualInflation ? annualInflation : 0
    const monthlyExpensesValue = showMonthlyExpenses ? monthlyExpenses : 0
    const rate = (annualInterestRate - annualInflationValue) / 100
    const totalSavings = (baseSalary - monthlyExpensesValue) * (1 + rate * years)
    return totalSavings
  }

  handleChange = name => event => {
    this.setState({ [name]: event.target.checked })
  }

  controlSwitches() {
    return (
      <FormGroup
        row
        style={{ width: '60%', padding: '20px', marginLeft: 'auto', marginRight: 'auto' }}
      >
        <FormControlLabel
          className="control-show-monthly-expenses"
          control={
            <Switch
              checked={this.state.showMonthlyExpenses}
              onChange={this.handleChange('showMonthlyExpenses')}
              value="checkedA"
              color="primary"
            />
          }
          label="Monthly expenses"
        />
        <FormControlLabel
          className="control-show-annual-inflation"
          control={
            <Switch
              checked={this.state.showAnnualInflation}
              onChange={this.handleChange('showAnnualInflation')}
              value="checkedB"
              color="primary"
            />
          }
          label="Annual Inflation"
        />
      </FormGroup>
    )
  }

  render() {
    const { showAnnualInflation, showMonthlyExpenses } = this.state
    return (
      <React.Fragment>
        <p>Calculate your savings</p>
        {this.controlSwitches()}
        <div className="savings-app-container">
          <Typography variant="body1" align="right" className="base-salary-label">
            Base salary
          </Typography>
          <InputFrame>
            <CurrencyInput
              placeholder="Base Salary"
              onValueChange={values => this.setState({ baseSalary: values })}
            />
          </InputFrame>
          {showMonthlyExpenses && (
            <React.Fragment>
              <Typography variant="body1" align="right">
                Monthly expenses
              </Typography>
              <CurrencyInput
                placeholder="Monthly expenses"
                onValueChange={values => this.setState({ monthlyExpenses: values })}
              />
            </React.Fragment>
          )}
          {showAnnualInflation && (
            <React.Fragment>
              <Typography variant="body1" align="right">
                Annual Inflation
              </Typography>
              <PercentageInput
                placeholder="Annual Inflation"
                onValueChange={values => this.setState({ annualInflation: values })}
              />
            </React.Fragment>
          )}
          <Typography variant="body1" align="right" className="annual-interest-rate-label">
            Annual Interest Rate
          </Typography>
          <PercentageInput
            placeholder="Annual Interest Rate"
            onValueChange={values => this.setState({ annualInterestRate: values })}
          />
          <Typography variant="body1" align="right" className="number-years-label">
            # years
          </Typography>
          <NumberFormat
            placeholder="years"
            onValueChange={values => this.setState({ years: values.floatValue })}
          />
          <Typography variant="body1" align="right">
            Annual savings
          </Typography>
          <NumberFormat thousandSeparator={true} prefix={'$'} value={this.state.totalSavings} />
        </div>
      </React.Fragment>
    )
  }
}

export default SavingsApp
