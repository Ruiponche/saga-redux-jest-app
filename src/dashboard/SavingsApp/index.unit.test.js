import React from 'react'
import renderer from 'react-test-renderer'
import SavingsApp from './SavingsApp'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

configure({ adapter: new Adapter() })

it('renders correctly', () => {
  const tree = renderer.create(<SavingsApp />).toJSON()
  expect(tree).toMatchSnapshot()
})

describe('<SavingsApp/>', () => {
  it('check presence of labels', () => {
    const wrapper = shallow(<SavingsApp />)
    expect(wrapper.contains([<p>Calculate your savings</p>])).toEqual(true)
    expect(wrapper.exists('.base-salary-label')).toEqual(true)
    expect(wrapper.exists('.annual-interest-rate-label')).toEqual(true)
    expect(wrapper.exists('.number-years-label')).toEqual(true)
  })
})
