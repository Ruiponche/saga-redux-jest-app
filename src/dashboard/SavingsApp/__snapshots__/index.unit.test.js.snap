// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <p>
    Calculate your savings
  </p>,
  <div
    className="MuiFormGroup-root-1 MuiFormGroup-row-2"
    style={
      Object {
        "marginLeft": "auto",
        "marginRight": "auto",
        "padding": "20px",
        "width": "60%",
      }
    }
  >
    <label
      className="MuiFormControlLabel-root-3 control-show-monthly-expenses"
    >
      <span
        className="MuiSwitch-root-9"
      >
        <span
          className="MuiButtonBase-root-28 MuiIconButton-root-22 MuiPrivateSwitchBase-root-18 MuiSwitch-switchBase-12 MuiSwitch-colorPrimary-14"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label-27"
          >
            <span
              className="MuiSwitch-icon-10"
            />
            <input
              checked={false}
              className="MuiPrivateSwitchBase-input-21"
              onChange={[Function]}
              type="checkbox"
              value="checkedA"
            />
          </span>
          <span
            className="MuiTouchRipple-root-66"
          />
        </span>
        <span
          className="MuiSwitch-bar-17"
        />
      </span>
      <span
        className="MuiTypography-root-31 MuiTypography-body1-40 MuiFormControlLabel-label-8"
      >
        Monthly expenses
      </span>
    </label>
    <label
      className="MuiFormControlLabel-root-3 control-show-annual-inflation"
    >
      <span
        className="MuiSwitch-root-9"
      >
        <span
          className="MuiButtonBase-root-28 MuiIconButton-root-22 MuiPrivateSwitchBase-root-18 MuiSwitch-switchBase-12 MuiSwitch-colorPrimary-14"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label-27"
          >
            <span
              className="MuiSwitch-icon-10"
            />
            <input
              checked={false}
              className="MuiPrivateSwitchBase-input-21"
              onChange={[Function]}
              type="checkbox"
              value="checkedB"
            />
          </span>
          <span
            className="MuiTouchRipple-root-66"
          />
        </span>
        <span
          className="MuiSwitch-bar-17"
        />
      </span>
      <span
        className="MuiTypography-root-31 MuiTypography-body1-40 MuiFormControlLabel-label-8"
      >
        Annual Inflation
      </span>
    </label>
  </div>,
  <div
    className="savings-app-container"
  >
    <p
      className="MuiTypography-root-31 MuiTypography-body1-40 MuiTypography-alignRight-55 base-salary-label"
    >
      Base salary
    </p>
    <div
      className="normal frame"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <input
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseUp={[Function]}
        placeholder="Base Salary"
        type="text"
        value=""
      />
    </div>
    <p
      className="MuiTypography-root-31 MuiTypography-body1-40 MuiTypography-alignRight-55 annual-interest-rate-label"
    >
      Annual Interest Rate
    </p>
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      placeholder="Annual Interest Rate"
      type="text"
      value=""
    />
    <p
      className="MuiTypography-root-31 MuiTypography-body1-40 MuiTypography-alignRight-55 number-years-label"
    >
      # years
    </p>
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      placeholder="years"
      type="text"
      value=""
    />
    <p
      className="MuiTypography-root-31 MuiTypography-body1-40 MuiTypography-alignRight-55"
    >
      Annual savings
    </p>
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      type="text"
      value=""
    />
  </div>,
]
`;
